canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+                    pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatency_datum, data->pDimension_per_stream, cwRequest);
+                    LOG_DEBUG("Received Ack Latencyy: " << receivedAckLatency);
+                    if (data->pCanaryConfig->useAggMetrics)
+                    {
+                        pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatency_datum, data->pAggregated_dimension, cwRequest);
+                    }
+
+                    auto outcome = data->pCWclient->PutMetricData(cwRequest);
+                    if (!outcome.IsSuccess())
+                    {
+                        cout << "Failed to put ReceivedAckLatency metric data:" << outcome.GetError().GetMessage() << endl;
+                    }
+                    else
+                    {
+                        cout << "Successfully put ReceivedAckLatency metric data" << endl;
+                    }
+                }
+        } else
+        {
+            cout << "Not sending Ack Latency metric because: timeOfFragmentEndSent < pFragmentAck->timestamp" << endl;
+        }
+    }
+}
+
+}  // namespace video
+}  // namespace kinesis
+}  // namespace amazonaws
+}  // namespace com;
+
+// add frame pts, frame index, original frame size, CRC to beginning of buffer
+VOID addCanaryMetadataToFrameData(PFrame pFrame)
+{
+    PBYTE pCurPtr = pFrame->frameData;
+    putUnalignedInt64BigEndian((PINT64) pCurPtr, pFrame->presentationTs / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+    pCurPtr += SIZEOF(UINT64);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->index);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->size);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, COMPUTE_CRC32(pFrame->frameData, pFrame->size));
+}
+
+VOID createCanaryFrameData(PFrame pFrame)
+{
+    UINT32 i;
+
+    for (i = CANARY_METADATA_SIZE; i < pFrame->size; i++) {
+        pFrame->frameData[i] = RAND();
+    }
+    addCanaryMetadataToFrameData(pFrame);
+}
+
+void create_kinesis_video_frame(Frame *frame, const nanoseconds &pts, const nanoseconds &dts, FRAME_FLAGS flags,
+                                void *data, size_t len) {
+    frame->flags = flags;
+    frame->decodingTs = static_cast<UINT64>(dts.count()) / DEFAULT_TIME_UNIT_IN_NANOS;
+    frame->presentationTs = static_cast<UINT64>(pts.count()) / DEFAULT_TIME_UNIT_IN_NANOS;
+    // set duration to 0 due to potential high spew from rtsp streams
+    frame->duration = 0;
+    frame->size = static_cast<UINT32>(len);
+    frame->frameData = reinterpret_cast<PBYTE>(data);
+    frame->trackId = DEFAULT_TRACK_ID;
+}
+
+void updateFragmentEndTimes(UINT64 curKeyFrameTime, uint64_t &lastKeyFrameTime, map<uint64_t, uint64_t> *mapPtr)
```

Comment:
Hmm. Doesnt make sense. Why do you need an address for lastKeyFrameTime? Why are you using address instead of pointer? It needs to be `PUINT64 lastKeyFrameTime`. And in the place where this is being called, you need to send address.

Reply:
Isn't this a reference? https://www.ibm.com/docs/en/zos/2.4.0?topic=calls-pass-by-reference-c-only